src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.System.IO.Pipelines;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
+{
+    public class RequestBodyReader : IRequestBodyReader
+    {
+        private readonly IPipe _pipe;
+        public RequestBodyReader(IPipe pipe)
+        {
+            _pipe = pipe;
+        }
+
+        public async Task StartAsync(MessageBody messageBody, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                while (true)
+                {
+                    var writableBuffer = _pipe.Writer.Alloc(1);
+                    int bytesRead;
+
+                    try
+                    {
+                        bytesRead = await messageBody.ReadAsync(writableBuffer.Buffer.GetArray(), cancellationToken);
+                        writableBuffer.Advance(bytesRead);
+                    }
+                    finally
+                    {
+                        writableBuffer.Commit();
+                    }
+
+                    if (bytesRead == 0)
```

Comment:
I prefer this pattern:

https://github.com/aspnet/KestrelHttpServer/blob/37f15bdd85cf9e32904232554202dc4a6052ef42/src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs#L87-L102